PLU-125: [EXCEL-7] Create table row action

ogp-weeloong opened the pull request:
## Problem
Users want to insert a row of data into their excel file.

## Solution
See our M365 context doc on notion for more details, but essentially, we instruct users to use Excel [tables](https://support.microsoft.com/en-au/office/overview-of-excel-tables-7ab0bb7d-3a9e-4b56-a3c9-6c94334e492c) so that we can use the [createTableRow](https://learn.microsoft.com/en-us/graph/api/table-post-rows?view=graph-rest-1.0&tabs=http) API to insert rows into a spreadsheet.

This is the only data insertion API I've found that doesn't use cell coordinates; this works around the "read empty row coordinates " race without us having to make entire actions atomic.

That said, for race-free behaviour, we still need to use the same session ID for each excel file across our fleet. I've created a new redis DB called `APP_DATA` to store each file's session ID, and used a distributed redis lock to manipulate this safely. See the giant comment block in the code for more details.

## Tests
- Concurrently run multiple createTableRow calls on the same session ID (via k6) and check that there is no data loss.
- Check that the table row gets inserted successfully by the code, and data is inserted in the correct columns.
- Check that if new columns are inserted in the table, existing pipe still inserts data in the correct column (matching by column name).
- Check that an error is thrown if user tries to update the same column in different MultiRow entries.
- Check that Graph API errors are correctly parsed, and that sessions are re-created if we ever get a `invalidSessionReCreatable` error code

**Unit tests to be added in a later PR if team agrees with approach.**

## Deploy Notes
**New dependencies**:
- `redlock`: distributed lock

ogp-weeloong commented on packages/backend/src/apps/m365-excel/common/retry-step.ts:
```diff
@@ -9,7 +9,7 @@ import { generateStepError } from '@/helpers/generate-step-error'
  * This is only used for non-fatal errors such as:
  * - Internal rate limiting
- * - er.... only internal rate limiting for now. <.<
+ * - Redis lock contention
```
Ignore this... i pretty much remove this error class later.

ogp-weeloong commented on packages/backend/src/apps/m365-excel/common/workbook-session/index.ts:
```diff
@@ -0,0 +1,306 @@
+  fileId: string,
+  $: IGlobalVariable,
+): Promise<string> {
+  return await runWithLock(tenant, fileId, async (signal) => {
```
For now, failing to acquire the lock will immediately fail the step; I will add retries in a later PR

m0nggh replied:
Is this what you are referring to? I thought session id expires in an hour?
<img width="904" alt="image" src="https://github.com/opengovsg/plumber/assets/65110268/30eac270-39e7-49d6-b104-7cd05a8fff94">

m0nggh replied:
does this get revalidated after an hour when the key expires? cus im manually deleting the key in redis to make it work

ogp-weeloong replied:
yes. Actually session ID expires in 5 minutes on Microsoft's end. Have changed the expiry timeout to match. I initially set expiry to an hour to ensure we would never start 2 sessions concurrently. But after discussion with ian, i think I was too paranoid.

m0nggh commented on packages/backend/src/apps/m365-excel/actions/create-table-row/index.ts:
```diff
@@ -0,0 +1,174 @@
+
+const action: IRawAction = {
+  name: 'Create row',
+  key: 'create-table-row',
```
nit: this should be camelCase?

pregnantboy replied:
yeah why u no stick to norm

ogp-weeloong replied:
old man forgot. Paiseh. Cos app keys are dash case I think.

m0nggh commented on packages/backend/src/apps/m365-excel/actions/create-table-row/index.ts:
```diff
@@ -0,0 +1,174 @@
+
+    $.setActionItem({
+      raw: {
+        relativeRowNumber: createRowResponse.data.index + 1,
```
nit: is tableRowNumber a clearer way to represent it? I was confused initially while testing

pregnantboy replied:
agreed, maybe sheetRowNumber and tableRowNumber?

ogp-weeloong replied:
good call, done!

m0nggh commented on packages/backend/src/apps/m365-excel/common/workbook-session/index.ts:
```diff
@@ -0,0 +1,306 @@
+  })
+}
+
+/**
```
wah this is super complex to me, would most probably need to kindly request some time from you to walkthrough and clear some doubts

ogp-weeloong replied:
ping me when you free, I can walk you through

pregnantboy commented on packages/backend/src/apps/m365-excel/dynamic-data/list-table-columns/index.ts:
```diff
@@ -0,0 +1,37 @@
+        name: `Column ${entry.index} (${entry.name})`,
+        // We return indices because excel works with indices, not names.
+        value: String(entry.index),
+      })),
```
this might be quite confusing to users

<img width="776" alt="image" src="https://github.com/opengovsg/plumber/assets/10072985/0fff6d69-4c4d-4932-9237-cd16e125e835">

suggest 
```javascript
name: entry.name,
value: `Column ${entry.index + 1}`
```
and then parse the value back to index later on

ogp-weeloong replied:
Thanks I was mentioning to stacey how this screenshot would give her heart attack

ogp-weeloong replied:
Did a little more digging and I discovered that I can match column name to its index without needing an additional query. Coded that up instead, seems more intuitive.

pregnantboy commented on packages/backend/src/apps/m365-excel/actions/create-table-row/index.ts:
```diff
@@ -0,0 +1,174 @@
+        relativeRowNumber: createRowResponse.data.index + 1,
+        rowNumber:
+          tableInfoResponse.data.rowIndex + createRowResponse.data.index + 2,
+        success: true,
```
nit: add some comments on why +1 or +2

ogp-weeloong replied:
done!

pregnantboy commented on packages/backend/src/apps/m365-excel/common/parse-graph-api-error.ts:
```diff
@@ -0,0 +1,33 @@
+): z.infer<typeof graphApiErrorSchema> | null {
+  try {
+    return graphApiErrorSchema.parse(
+      JSON.parse(String(err.response?.data ?? '')),
```
String(err.response?.data) will always give u '[object Object]'. why are u attempting to stringify and then JSON.parse again?

ogp-weeloong replied:
I derped. late night coding and didnt test this bit properly. paiseh >.<. Fixed

pregnantboy commented on packages/backend/src/apps/m365-excel/common/workbook-session/redis.ts:
```diff
@@ -0,0 +1,113 @@
+// Session ID redis key expiry
+//
+// Arbitrarily set to 1 hour. See giant comment in index.ts for more details.
+const SESSION_ID_EXPIRY_SECONDS = 60 * 60
```
ref: https://learn.microsoft.com/en-us/graph/api/resources/excel?view=graph-rest-1.0#usage

> Typically the persistent session expires after about 5 minutes of inactivity. Non persistent session expires after about 7 minutes of inactivity.

pregnantboy reviewed (changes requested):
session id not refreshing due to incorrect error parsing. also, is 1 hour session id expiry (instead of 5 mins) intentional?

pregnantboy commented on packages/backend/src/apps/m365-excel/common/workbook-session/redis.ts:
```diff
@@ -0,0 +1,113 @@
+): Promise<string | null> {
+  const key = `${makeRedisKeyPrefix(tenant, fileId)}session:id`
+
+  const [[hgetErr, sessionId], [expireErr]] = await redisAppDataClient
```
```suggestion
  const [[getErr, sessionId], [expireErr]] = await redisAppDataClient
```

does await throw the error or return it as a response here?

ogp-weeloong replied:
await returns the error in the response (e.g. to test, change the expire command to an invalid time string - e.g. `10o`.)

According to this old ioredis [discussion](https://github.com/redis/ioredis/issues/301#issuecomment-218641172), It looks like ioredis has "split" error handling:
- outright syntax errors -> throw error
- command execution error -> return error

One of the contributions mentioned wanting to change this in ioredis v3, but as far as I can tell this double-error handling status quo has remained.

pregnantboy commented on packages/backend/src/apps/m365-excel/common/workbook-session/index.ts:
```diff
@@ -0,0 +1,306 @@
+      if (sessionIdInRedis !== this.sessionId) {
+        throw new Error(
+          "M365 servers encountered an error and invalidated the current Excel session; please double check the file's data",
+        )
```
other reasons for this error could be redis related (e.g. failed to write to redis, failed to read from redis, data was deleted..)

ogp-weeloong replied:
Have word-smithed the error message a little bit

ogp-weeloong commented:
Addressed comments, with one more further change: i discovered I can use another API to match column name to the column index efficiently, so I coded that up instead.

Further tests done:
- Check that the table row gets inserted successfully by the code, and data is inserted in the correct columns.
- Check that if new columns are inserted in the table, existing pipe still inserts data in the correct column (matching by column name).
- Check that an error is thrown if user tries to update the same column in different MultiRow entries.
- Check that Graph API errors are correctly parsed, and that sessions are re-created if we ever get a `invalidSessionReCreatable` error code

pregnantboy commented on packages/backend/src/apps/m365-excel/actions/create-table-row/index.ts:
```diff
@@ -135,12 +145,36 @@ const action: IRawAction = {
+      if (seenColumnNames.has(currColumnName)) {
+        throw new Error(
+          `Trying to write 2 different values to the same column (${currColumnName})`,
+        )
```
nit: use `generateStepError` to nicely return this error

pregnantboy replied:
edit: use new StepError once malcolm's pr is merged

ogp-weeloong replied:
Thanks! was planning to convert all the acitons to generateStepError in a later PR to avoid rebase hell. but might as well do it now, I think no impact on rebasing.

m0nggh replied:
i was thinking you can merge ur excel PRs first and i can help to resolve the merge conflict on my PR later on

ogp-weeloong replied:
nvm i refactored already no worries

pregnantboy reviewed (approved):
lgtm
